{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "import numpy as np\n",
    "\n",
    "from IPython.display import clear_output, Image, display, HTML\n",
    "from tensorflow.examples.tutorials.mnist import input_data #import input_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting ../hands_on/data/train-images-idx3-ubyte.gz\n",
      "Extracting ../hands_on/data/train-labels-idx1-ubyte.gz\n",
      "Extracting ../hands_on/data/t10k-images-idx3-ubyte.gz\n",
      "Extracting ../hands_on/data/t10k-labels-idx1-ubyte.gz\n"
     ]
    }
   ],
   "source": [
    "#################################################################################\n",
    "'''\n",
    "Data is downloaded as an mnist object\n",
    "At every iteration, the batch of images are read from the dataset \n",
    "'''\n",
    "#################################################################################\n",
    "batch_size = 100\n",
    "mnist = input_data.read_data_sets('../hands_on/data', one_hot=True)\n",
    "def get_mnist_batch(mnist, train):\n",
    "    if train:\n",
    "        xs, ys = mnist.train.next_batch(batch_size)\n",
    "        #k = FLAGS.dropout\n",
    "    else:\n",
    "        xs, ys = mnist.test.next_batch(batch_size)\n",
    "        #k = 1.0\n",
    "    return (2*xs)-1, ys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#################################################################################\n",
    "'''\n",
    "This function creates the neural network with layer definitions\n",
    "Input: Images as tensors \n",
    "Output: 10-d tensor for each image\n",
    "'''\n",
    "#################################################################################\n",
    "def neural_network(ip):\n",
    "    # Define the layers for your neural network \n",
    "    hidden_layer1 = tf.layers.dense(ip, 1296, activation=tf.nn.relu, name='hidden_layer1') \n",
    "    hidden_layer2 = tf.layers.dense(hidden_layer1, 1296, activation=tf.nn.relu, name='hidden_layer2') \n",
    "    hidden_layer3 = tf.layers.dense(hidden_layer2, 1296, activation=tf.nn.relu, name='hidden_layer3') \n",
    "    op = tf.layers.dense(hidden_layer3, 10, activation=tf.nn.relu, name='output') \n",
    "    return op\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#################################################################################\n",
    "'''\n",
    "Create Graph first by defining placeholders, ops, loss functions and optimizers\n",
    "\n",
    "'''\n",
    "#################################################################################\n",
    "\n",
    "tf.reset_default_graph()\n",
    "#################################################################################\n",
    "'''Placeholders'''\n",
    "#################################################################################\n",
    "with tf.variable_scope('inputs'):\n",
    "    inputs = tf.placeholder(tf.float32, shape=[None,784])\n",
    "with tf.variable_scope('ground_truth'):\n",
    "    ground_truth= tf.placeholder(tf.float32, shape=[None,10])\n",
    "\n",
    "#################################################################################\n",
    "'''MNIST Neural Network'''\n",
    "#################################################################################\n",
    "with tf.variable_scope('mnist_network'):\n",
    "    outputs = neural_network(inputs)\n",
    "\n",
    "    \n",
    "#################################################################################\n",
    "'''Loss and Optimizers'''\n",
    "#################################################################################\n",
    "learning_rate = 0.0001\n",
    "with tf.variable_scope('cross_entropy_loss'):\n",
    "    # Define the loss between the logits and the labels\n",
    "    loss = tf.nn.softmax_cross_entropy_with_logits(labels=ground_truth, logits=outputs)\n",
    "    \n",
    "    # Define the optimizer and minimize the loss using it\n",
    "    optimizer = tf.train.AdamOptimizer(learning_rate).minimize(loss)\n",
    "\n",
    "accuracy = []\n",
    "with tf.name_scope('accuracy'):\n",
    "    accuracy = tf.reduce_mean(tf.cast(tf.equal(tf.argmax(outputs, 1), tf.argmax(ground_truth, 1)), tf.float32))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy at step 0: 0.07\n",
      "Accuracy at step 10: 0.4\n",
      "Accuracy at step 20: 0.42\n",
      "Accuracy at step 30: 0.46\n",
      "Accuracy at step 40: 0.55\n",
      "Accuracy at step 50: 0.48\n",
      "Accuracy at step 60: 0.51\n",
      "Accuracy at step 70: 0.56\n",
      "Accuracy at step 80: 0.49\n",
      "Accuracy at step 90: 0.58\n",
      "Accuracy at step 100: 0.55\n",
      "Accuracy at step 110: 0.54\n",
      "Accuracy at step 120: 0.48\n",
      "Accuracy at step 130: 0.61\n",
      "Accuracy at step 140: 0.58\n",
      "Accuracy at step 150: 0.52\n",
      "Accuracy at step 160: 0.56\n",
      "Accuracy at step 170: 0.58\n",
      "Accuracy at step 180: 0.54\n",
      "Accuracy at step 190: 0.55\n",
      "Accuracy at step 200: 0.55\n",
      "Accuracy at step 210: 0.55\n",
      "Accuracy at step 220: 0.63\n",
      "Accuracy at step 230: 0.58\n",
      "Accuracy at step 240: 0.62\n",
      "Accuracy at step 250: 0.54\n",
      "Accuracy at step 260: 0.64\n",
      "Accuracy at step 270: 0.57\n",
      "Accuracy at step 280: 0.53\n",
      "Accuracy at step 290: 0.56\n",
      "Accuracy at step 300: 0.55\n",
      "Accuracy at step 310: 0.57\n",
      "Accuracy at step 320: 0.56\n",
      "Accuracy at step 330: 0.57\n",
      "Accuracy at step 340: 0.58\n",
      "Accuracy at step 350: 0.61\n",
      "Accuracy at step 360: 0.57\n",
      "Accuracy at step 370: 0.56\n",
      "Accuracy at step 380: 0.62\n",
      "Accuracy at step 390: 0.61\n",
      "Accuracy at step 400: 0.55\n",
      "Accuracy at step 410: 0.54\n",
      "Accuracy at step 420: 0.55\n",
      "Accuracy at step 430: 0.58\n",
      "Accuracy at step 440: 0.6\n",
      "Accuracy at step 450: 0.58\n",
      "Accuracy at step 460: 0.61\n",
      "Accuracy at step 470: 0.66\n",
      "Accuracy at step 480: 0.62\n",
      "Accuracy at step 490: 0.67\n",
      "Accuracy at step 500: 0.57\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Parent directory of my_mnist_model doesn't exist, can't save.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNotFoundError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1322\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1323\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1324\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mOpError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_run_fn\u001b[0;34m(session, feed_dict, fetch_list, target_list, options, run_metadata)\u001b[0m\n\u001b[1;32m   1301\u001b[0m                                    \u001b[0mfeed_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfetch_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtarget_list\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1302\u001b[0;31m                                    status, run_metadata)\n\u001b[0m\u001b[1;32m   1303\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/errors_impl.py\u001b[0m in \u001b[0;36m__exit__\u001b[0;34m(self, type_arg, value_arg, traceback_arg)\u001b[0m\n\u001b[1;32m    472\u001b[0m             \u001b[0mcompat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mas_text\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc_api\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTF_Message\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 473\u001b[0;31m             c_api.TF_GetCode(self.status.status))\n\u001b[0m\u001b[1;32m    474\u001b[0m     \u001b[0;31m# Delete the underlying status object from memory otherwise it stays alive\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNotFoundError\u001b[0m: ; No such file or directory\n\t [[Node: save/SaveV2 = SaveV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](_arg_save/Const_0_0, save/SaveV2/tensor_names, save/SaveV2/shape_and_slices, cross_entropy_loss/beta1_power, cross_entropy_loss/beta2_power, cross_entropy_loss/mnist_network/hidden_layer1/bias/Adam, cross_entropy_loss/mnist_network/hidden_layer1/bias/Adam_1, cross_entropy_loss/mnist_network/hidden_layer1/kernel/Adam, cross_entropy_loss/mnist_network/hidden_layer1/kernel/Adam_1, cross_entropy_loss/mnist_network/hidden_layer2/bias/Adam, cross_entropy_loss/mnist_network/hidden_layer2/bias/Adam_1, cross_entropy_loss/mnist_network/hidden_layer2/kernel/Adam, cross_entropy_loss/mnist_network/hidden_layer2/kernel/Adam_1, cross_entropy_loss/mnist_network/hidden_layer3/bias/Adam, cross_entropy_loss/mnist_network/hidden_layer3/bias/Adam_1, cross_entropy_loss/mnist_network/hidden_layer3/kernel/Adam, cross_entropy_loss/mnist_network/hidden_layer3/kernel/Adam_1, cross_entropy_loss/mnist_network/output/bias/Adam, cross_entropy_loss/mnist_network/output/bias/Adam_1, cross_entropy_loss/mnist_network/output/kernel/Adam, cross_entropy_loss/mnist_network/output/kernel/Adam_1, mnist_network/hidden_layer1/bias, mnist_network/hidden_layer1/kernel, mnist_network/hidden_layer2/bias, mnist_network/hidden_layer2/kernel, mnist_network/hidden_layer3/bias, mnist_network/hidden_layer3/kernel, mnist_network/output/bias, mnist_network/output/kernel)]]",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mNotFoundError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/tensorflow/python/training/saver.py\u001b[0m in \u001b[0;36msave\u001b[0;34m(self, sess, save_path, global_step, latest_filename, meta_graph_suffix, write_meta_graph, write_state)\u001b[0m\n\u001b[1;32m   1572\u001b[0m               \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msaver_def\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave_tensor_name\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1573\u001b[0;31m               {self.saver_def.filename_tensor_name: checkpoint_file})\n\u001b[0m\u001b[1;32m   1574\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m    888\u001b[0m       result = self._run(None, fetches, feed_dict, options_ptr,\n\u001b[0;32m--> 889\u001b[0;31m                          run_metadata_ptr)\n\u001b[0m\u001b[1;32m    890\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0mrun_metadata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_run\u001b[0;34m(self, handle, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1119\u001b[0m       results = self._do_run(handle, final_targets, final_fetches,\n\u001b[0;32m-> 1120\u001b[0;31m                              feed_dict_tensor, options, run_metadata)\n\u001b[0m\u001b[1;32m   1121\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_run\u001b[0;34m(self, handle, target_list, fetch_list, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1316\u001b[0m       return self._do_call(_run_fn, self._session, feeds, fetches, targets,\n\u001b[0;32m-> 1317\u001b[0;31m                            options, run_metadata)\n\u001b[0m\u001b[1;32m   1318\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1335\u001b[0m           \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1336\u001b[0;31m       \u001b[0;32mraise\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode_def\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmessage\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1337\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNotFoundError\u001b[0m: ; No such file or directory\n\t [[Node: save/SaveV2 = SaveV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](_arg_save/Const_0_0, save/SaveV2/tensor_names, save/SaveV2/shape_and_slices, cross_entropy_loss/beta1_power, cross_entropy_loss/beta2_power, cross_entropy_loss/mnist_network/hidden_layer1/bias/Adam, cross_entropy_loss/mnist_network/hidden_layer1/bias/Adam_1, cross_entropy_loss/mnist_network/hidden_layer1/kernel/Adam, cross_entropy_loss/mnist_network/hidden_layer1/kernel/Adam_1, cross_entropy_loss/mnist_network/hidden_layer2/bias/Adam, cross_entropy_loss/mnist_network/hidden_layer2/bias/Adam_1, cross_entropy_loss/mnist_network/hidden_layer2/kernel/Adam, cross_entropy_loss/mnist_network/hidden_layer2/kernel/Adam_1, cross_entropy_loss/mnist_network/hidden_layer3/bias/Adam, cross_entropy_loss/mnist_network/hidden_layer3/bias/Adam_1, cross_entropy_loss/mnist_network/hidden_layer3/kernel/Adam, cross_entropy_loss/mnist_network/hidden_layer3/kernel/Adam_1, cross_entropy_loss/mnist_network/output/bias/Adam, cross_entropy_loss/mnist_network/output/bias/Adam_1, cross_entropy_loss/mnist_network/output/kernel/Adam, cross_entropy_loss/mnist_network/output/kernel/Adam_1, mnist_network/hidden_layer1/bias, mnist_network/hidden_layer1/kernel, mnist_network/hidden_layer2/bias, mnist_network/hidden_layer2/kernel, mnist_network/hidden_layer3/bias, mnist_network/hidden_layer3/kernel, mnist_network/output/bias, mnist_network/output/kernel)]]\n\nCaused by op 'save/SaveV2', defined at:\n  File \"/usr/lib/python3.4/runpy.py\", line 170, in _run_module_as_main\n    \"__main__\", mod_spec)\n  File \"/usr/lib/python3.4/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/usr/local/lib/python3.4/dist-packages/ipykernel_launcher.py\", line 16, in <module>\n    app.launch_new_instance()\n  File \"/usr/local/lib/python3.4/dist-packages/traitlets/config/application.py\", line 658, in launch_instance\n    app.start()\n  File \"/usr/local/lib/python3.4/dist-packages/ipykernel/kernelapp.py\", line 477, in start\n    ioloop.IOLoop.instance().start()\n  File \"/usr/lib/python3/dist-packages/zmq/eventloop/ioloop.py\", line 160, in start\n    super(ZMQIOLoop, self).start()\n  File \"/usr/local/lib/python3.4/dist-packages/tornado/ioloop.py\", line 888, in start\n    handler_func(fd_obj, events)\n  File \"/usr/local/lib/python3.4/dist-packages/tornado/stack_context.py\", line 277, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/usr/lib/python3/dist-packages/zmq/eventloop/zmqstream.py\", line 433, in _handle_events\n    self._handle_recv()\n  File \"/usr/lib/python3/dist-packages/zmq/eventloop/zmqstream.py\", line 465, in _handle_recv\n    self._run_callback(callback, msg)\n  File \"/usr/lib/python3/dist-packages/zmq/eventloop/zmqstream.py\", line 407, in _run_callback\n    callback(*args, **kwargs)\n  File \"/usr/local/lib/python3.4/dist-packages/tornado/stack_context.py\", line 277, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/usr/local/lib/python3.4/dist-packages/ipykernel/kernelbase.py\", line 283, in dispatcher\n    return self.dispatch_shell(stream, msg)\n  File \"/usr/local/lib/python3.4/dist-packages/ipykernel/kernelbase.py\", line 235, in dispatch_shell\n    handler(stream, idents, msg)\n  File \"/usr/local/lib/python3.4/dist-packages/ipykernel/kernelbase.py\", line 399, in execute_request\n    user_expressions, allow_stdin)\n  File \"/usr/local/lib/python3.4/dist-packages/ipykernel/ipkernel.py\", line 196, in do_execute\n    res = shell.run_cell(code, store_history=store_history, silent=silent)\n  File \"/usr/local/lib/python3.4/dist-packages/ipykernel/zmqshell.py\", line 533, in run_cell\n    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n  File \"/usr/local/lib/python3.4/dist-packages/IPython/core/interactiveshell.py\", line 2728, in run_cell\n    interactivity=interactivity, compiler=compiler, result=result)\n  File \"/usr/local/lib/python3.4/dist-packages/IPython/core/interactiveshell.py\", line 2850, in run_ast_nodes\n    if self.run_code(code, result):\n  File \"/usr/local/lib/python3.4/dist-packages/IPython/core/interactiveshell.py\", line 2910, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"<ipython-input-5-81b34504aebe>\", line 36, in <module>\n    saver = tf.train.Saver()\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/training/saver.py\", line 1218, in __init__\n    self.build()\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/training/saver.py\", line 1227, in build\n    self._build(self._filename, build_save=True, build_restore=True)\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/training/saver.py\", line 1263, in _build\n    build_save=build_save, build_restore=build_restore)\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/training/saver.py\", line 748, in _build_internal\n    save_tensor = self._AddSaveOps(filename_tensor, saveables)\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/training/saver.py\", line 296, in _AddSaveOps\n    save = self.save_op(filename_tensor, saveables)\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/training/saver.py\", line 239, in save_op\n    tensors)\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/ops/gen_io_ops.py\", line 1163, in save_v2\n    shape_and_slices=shape_and_slices, tensors=tensors, name=name)\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/op_def_library.py\", line 787, in _apply_op_helper\n    op_def=op_def)\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/ops.py\", line 2956, in create_op\n    op_def=op_def)\n  File \"/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/ops.py\", line 1470, in __init__\n    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access\n\nNotFoundError (see above for traceback): ; No such file or directory\n\t [[Node: save/SaveV2 = SaveV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](_arg_save/Const_0_0, save/SaveV2/tensor_names, save/SaveV2/shape_and_slices, cross_entropy_loss/beta1_power, cross_entropy_loss/beta2_power, cross_entropy_loss/mnist_network/hidden_layer1/bias/Adam, cross_entropy_loss/mnist_network/hidden_layer1/bias/Adam_1, cross_entropy_loss/mnist_network/hidden_layer1/kernel/Adam, cross_entropy_loss/mnist_network/hidden_layer1/kernel/Adam_1, cross_entropy_loss/mnist_network/hidden_layer2/bias/Adam, cross_entropy_loss/mnist_network/hidden_layer2/bias/Adam_1, cross_entropy_loss/mnist_network/hidden_layer2/kernel/Adam, cross_entropy_loss/mnist_network/hidden_layer2/kernel/Adam_1, cross_entropy_loss/mnist_network/hidden_layer3/bias/Adam, cross_entropy_loss/mnist_network/hidden_layer3/bias/Adam_1, cross_entropy_loss/mnist_network/hidden_layer3/kernel/Adam, cross_entropy_loss/mnist_network/hidden_layer3/kernel/Adam_1, cross_entropy_loss/mnist_network/output/bias/Adam, cross_entropy_loss/mnist_network/output/bias/Adam_1, cross_entropy_loss/mnist_network/output/kernel/Adam, cross_entropy_loss/mnist_network/output/kernel/Adam_1, mnist_network/hidden_layer1/bias, mnist_network/hidden_layer1/kernel, mnist_network/hidden_layer2/bias, mnist_network/hidden_layer2/kernel, mnist_network/hidden_layer3/bias, mnist_network/hidden_layer3/kernel, mnist_network/output/bias, mnist_network/output/kernel)]]\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-81b34504aebe>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mExists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'my_mnist_model'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m         \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMakeDirs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'my_mnist_model'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m     \u001b[0msaver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msess\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'my_mnist_model'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.4/dist-packages/tensorflow/python/training/saver.py\u001b[0m in \u001b[0;36msave\u001b[0;34m(self, sess, save_path, global_step, latest_filename, meta_graph_suffix, write_meta_graph, write_state)\u001b[0m\n\u001b[1;32m   1592\u001b[0m               \"Parent directory of {} doesn't exist, can't save.\".format(\n\u001b[1;32m   1593\u001b[0m                   save_path))\n\u001b[0;32m-> 1594\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1595\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1596\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mwrite_meta_graph\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Parent directory of my_mnist_model doesn't exist, can't save."
     ]
    }
   ],
   "source": [
    "#################################################################################\n",
    "'''\n",
    "Create a TF session and execute the graph\n",
    "\n",
    "'''\n",
    "#################################################################################\n",
    "\n",
    "with tf.Session() as sess:\n",
    "    # initialize all the variables - this assigns the values to the variables\n",
    "    tf.global_variables_initializer().run()\n",
    "    \n",
    "    max_steps = 101\n",
    "    for i in range(max_steps):\n",
    "        # read data from mnist dataset and get a batch\n",
    "        input_, gt_ = get_mnist_batch(mnist, train=True)\n",
    "    \n",
    "        # The dictionary assigns the numerical values to the placeholders\n",
    "        values_dict = {inputs:input_, ground_truth:gt_}\n",
    "        fetch_values = [optimizer, loss, accuracy]\n",
    "        #################################################################################\n",
    "        '''\n",
    "        Run a session to execute the graph\n",
    "        sess.run() should consist of 'fetch' and 'feed' \n",
    "        fetch: are the list of variables to be returned\n",
    "        feed: is the dictionary of values assigned to respective placeholders\n",
    "        '''\n",
    "        #################################################################################\n",
    "        _, loss_, accuracy_ = sess.run(fetch_values, feed_dict=values_dict)\n",
    "        \n",
    "        \n",
    "        if i%10==0:\n",
    "            print(\"Accuracy at step %d: %g\"%(i, accuracy_))\n",
    "            #plt.imshow(input_[0].reshape([28,28]));plt.show()\n",
    "            \n",
    "    # Define the saver and save your model \n",
    "    saver = tf.train.Saver()\n",
    "    # save the model\n",
    "    if not tf.gfile.Exists('my_mnist_model'):\n",
    "        tf.gfile.MakeDirs('my_mnist_model')\n",
    "    saver.save(sess, 'my_mnist_model/model.ckpt')   \n",
    "    \n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
